starlite/contrib/repository/abc.py

```diff
@@ -0,0 +1,244 @@
+"""Data persistence interface."""
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from typing import TYPE_CHECKING, Any, Generic, TypeVar
+
+from .exceptions import NotFoundError
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from .types import FilterTypes
+
+
+__all__ = ("AbstractRepository",)
+
+T = TypeVar("T")
+RepoT = TypeVar("RepoT", bound="AbstractRepository")
+CollectionT = TypeVar("CollectionT")
+
+
+class AbstractRepository(Generic[T], metaclass=ABCMeta):
+    """Interface for persistent data interaction."""
+
+    model_type: type[T]
+    """Type of object represented by the repository."""
+    id_attribute = "id"
+    """Name of the primary identifying attribute on `model_type`."""
+
+    def __init__(self, **kwargs: Any) -> None:
+        """Repository constructors accept arbitrary kwargs."""
+        super().__init__(**kwargs)
+
+    @abstractmethod
+    async def add(self, data: T) -> T:
+        """Add `data` to the collection.
+
+        Args:
+            data: Instance to be added to the collection.
+
+        Returns:
+            The added instance.
+        """
+
+    @abstractmethod
+    async def add_many(self, data: Sequence[T]) -> Sequence[T]:
```

Comment:
does this actually need to be a sequence, or could it be any iterable, e.g., a set?

Reply:
I think `Iterable` should be fine IMO

Reply:
can you guys think of a way we can preserve the typing and combine add_many and add into a single endpoint?  There's no reason why we can't just use `add_many` for all inserts.

Reply:
I like the explicitness of separate methods.. you could use `@overload` though..

Reply:
Agree with @peterschutt on this. Two separate methods are better than one that does both.

Reply:
Fine with me.  Thanks for the quick feedback

Reply:
I could not make this work with an `Iterable`.  There are other places where we need an indexable iterator.  I've moved it back to a `Sequence` for now.

Reply:
I actually updated it to a `list` since that's what is returned.  Any issues with that type?

Reply:
> There are other places where we need an indexable iterator.

Such as?

We should always prefer abstract function inputs IMO, it allows for much nicer DX. E.g., `Iterable[ModelT]` would allow for things like:

```python
repo.add_many(Model(id=i) for i in ids)
```
or

```python
models = {i: Model(id=i) for id in ids}
repo.add_many(models.values())
```

Reply:
For instance, when I swap from list to `Iterable`, one of the method that begins complaining is `count`
```python
    async def count(self, *filters: FilterTypes, **kwargs: Any) -> int:
        return len(await self.list(*filters, **kwargs))
```
```Argument 1 to "len" has incompatible type "Iterable[ModelT]"; expected "Sized"  [arg-type]mypy(error)```

and 

```python
    async def get_one_or_none(self, **kwargs: Any) -> ModelT | None:
        data = await self.list(**kwargs)
        return data[0] if len(data) > 0 else None

```
gives
```Value of type "Iterable[ModelT]" is not indexable  [index]mypy(error)```

Reply:
This is in `GenericMockRepository`